src/stan_math_backend/Transform_Mir.ml

```diff
@@ -161,56 +161,97 @@ let data_read smeta (decl_id, st) =
-    , Program.({ out_constrained_st= cst
-               ; out_unconstrained_st= ucst
-               ; out_block; _ }) ) =
+    (decl_id, Program.({out_constrained_st= cst; out_block; out_trans; _})) =
   if not (out_block = Parameters) then []
   else
-    let decl_id, decl =
-      match cst = ucst with
-      | true -> (decl_id, [])
-      | false ->
-          let decl_id = decl_id ^ "_in__" in
-          let d =
-            Stmt.Fixed.Pattern.Decl
-              {decl_adtype= AutoDiffable; decl_id; decl_type= Sized ucst}
-          in
-          (decl_id, [Stmt.Fixed.{meta= smeta; pattern= d}])
+    let ut = SizedType.to_unsized cst in
+    let decl_var =
+      Expr.Fixed.
+        { pattern= Var decl_id
+        ; meta=
+            Expr.Typed.Meta.create ~loc:smeta ~type_:ut ~adlevel:AutoDiffable
+              () }
     in
-    let unconstrained_decl_var =
-      let meta =
-        Expr.Typed.Meta.create ~loc:smeta
-          ~type_:SizedType.(to_unsized cst)
-          ~adlevel:AutoDiffable ()
-      in
-      Expr.Fixed.{meta; pattern= Var decl_id}
+    let transform_args = transform_args out_trans in
+    (*
+      For constrains that return square / lower triangular matrices the C++
+      only wants one of the matrix dimensions.
+    *)
+    let rec constrain_get_dims st =
+      match st with
+      | SizedType.SInt | SReal -> []
+      | SVector d | SRowVector d -> [d]
+      | SMatrix (_, dim2) -> [dim2]
+      | SArray (t, dim) -> dim :: constrain_get_dims t
     in
-    let bodyfn var =
-      let readfnapp (var : Expr.Typed.t) =
-        Expr.(
-          Helpers.(
-            internal_funapp FnReadParam
-              ( str (base_ut_to_string (SizedType.to_unsized ucst))
-              :: SizedType.dims_of ucst ))
-            Typed.Meta.{var.meta with type_= base_type ucst})
-      in
-      Stmt.Helpers.assign_indexed (SizedType.to_unsized cst) decl_id smeta
-        readfnapp var
+    let read_constrain_dims constrain_transform st =
+      match constrain_transform with
+      | Program.CholeskyCorr | Correlation | Covariance ->
+          constrain_get_dims st
+      | _ -> SizedType.get_dims st
+    in
+    (* this is an absolute hack
+
+       I need to unpack the constraint arguments and the dimensions in codegen, but we pack them all together into a fake function FnReadParam as expressions
+       So, I'm packing in the number of constraint arguments to read as an int expression
+       To avoid this hack, keep internal functions as a variant type instead of a normal funapp
+    *)
+    let n_args_expression = Expr.Helpers.int (List.length transform_args) in
+    let read =
+      Expr.(
+        Helpers.(
+          internal_funapp FnReadParam
+            ( Expr.Helpers.str
+                (constrain_constraint_to_string out_trans Constrain)
+            :: n_args_expression
+            :: (transform_args @ read_constrain_dims out_trans cst) ))
+          Typed.Meta.{decl_var.meta with type_= ut})
```

Comment:
Would you mind adding an expect test for this Transform_mir step? I just realized it's only kind of implicitly tested and I'd really love to see the actual reified MIR before and after to help me see how this step works. It doesn't need to be super complicated but it'd be good if it had a constrained parameter. A thing that would be great would be picking some example model and adding a test that just runs the compiler up to right before the Transform_mir step, prints that out, then runs Transform_mir and prints it out again.

Reply:
This should be much simpler now, the MIR arguments just reflect how they're called in codegen. It's lousy that I added a an lp__ variable in the MIR but it's better than it was.